File: trunk/theLink/example/python/Filter4.py
#+
#§  \file       theLink/example/python/Filter4.py
#§  \brief      \$Id: Filter4.py 25 2009-12-04 11:05:46Z aotto1968 $
#§  
#§  (C) 2009 - NHI - #1 - Project - Group
#§  
#§  \version    \$Rev: 25 $
#§  \author     EMail: aotto1968 at users.berlios.de
#§  \attention  this software has GPL permissions to copy
#§              please contact AUTHORS for additional information
#§

import sys
from pymsgque import *

class Filter4(MqS):
  def __init__(ctx, tmpl=None):
    ctx.ConfigSetIgnoreExit(True)
    ctx.ConfigSetServerSetup(ctx.ServerSetup)
    ctx.ConfigSetServerCleanup(ctx.ServerCleanup)
    ctx.ConfigSetEvent(ctx.Event)
    ctx.itms = []
    ctx.FH = None
    MqS.__init__(ctx)
  def ServerCleanup(ctx):
    if ctx.FH != None:
      ctx.FH.close()
  def ServerSetup(ctx):
    ftr = ctx.ServiceGetFilter()
    ctx.ServiceCreate("LOGF", ctx.LOGF)
    ctx.ServiceCreate("EXIT", ctx.EXIT)
    ctx.ServiceCreate("+ALL", ctx.FilterIn)
    ftr.ServiceCreate("WRIT", ftr.WRIT)
    ftr.ServiceProxy("+TRT")
  def ErrorWrite(ctx):
    ctx.FH.write("ERROR: " + ctx.ErrorGetText() + "\n")
    ctx.FH.flush()
    ctx.ErrorReset()
  def LOGF(ctx):
    ftr = ctx.ServiceGetFilter()
    if ftr.LinkGetTargetIdent() == "transFilter":
      ctx.ReadForward(ftr)
    else:
      ctx.FH = open(ctx.ReadC(), "a")
    ctx.SendRETURN()
  def WRIT(ftr):
    FH = ftr.ServiceGetFilter().FH
    FH.write(ftr.ReadC() + "\n")
    FH.flush()
    ftr.SendRETURN()
  def EXIT(ctx):
    os._exit(1)
  def Event(ctx):
    if len(ctx.itms) == 0:
      ctx.ErrorSetCONTINUE()
    else:
      try:
        ftr = ctx.ServiceGetFilter()
        ftr.LinkConnect()
        ctx.ReadForward(ftr,ctx.itms[0])
      except:
        ctx.ErrorSet()
        if ctx.ErrorIsEXIT():
          ctx.ErrorReset()
          return
        else:
          ctx.ErrorWrite()
      ctx.itms.pop(0)
  def FilterIn(ctx):
    ctx.itms.append(ctx.ReadDUMP())
    ctx.SendRETURN()

FactoryDefault("transFilter", Filter4)
srv = Filter4()
try:
  srv.LinkCreate(sys.argv)
  srv.ProcessEvent(MqS_WAIT_FOREVER)
except:
  srv.ErrorSet()
finally:
  srv.Exit()

File: trunk/theLink/example/ruby/Filter4.rb
#+
#:  \file       theLink/example/ruby/Filter4.rb
#:  \brief      \$Id$
#:  
#:  (C) 2010 - NHI - #1 - Project - Group
#:  
#:  \version    \$Rev$
#:  \author     EMail: aotto1968 at users.berlios.de
#:  \attention  this software has GPL permissions to copy
#:              please contact AUTHORS for additional information
#:

require "rubymsgque"

class Filter4 < MqS
  attr_accessor :fh
  def initialize(tmpl=nil)
    super(tmpl)
    ConfigSetIgnoreExit(true)
    ConfigSetServerSetup(method(:ServerSetup))
    ConfigSetServerCleanup(method(:ServerCleanup))
    ConfigSetEvent(method(:Event))
    @itms = []
    @fh = nil
  end
  def ServerCleanup
    @fh.close if @fh != nil
    @fh = nil
  end
  def ServerSetup
    ftr = ServiceGetFilter()
    ServiceCreate("LOGF", method(:LOGF))
    ServiceCreate("EXIT", method(:EXIT))
    ServiceCreate("+ALL", method(:FilterIn))
    ftr.ServiceCreate("WRIT", ftr.method(:WRIT))
    ftr.ServiceProxy("+TRT")
  end
  def WRIT
    fh = ServiceGetFilter().fh
    fh.write(ReadC() + "\n")
    fh.flush()
    SendRETURN()
  end
  def ErrorWrite
    @fh.write("ERROR: " + ErrorGetText() + "\n")
    @fh.flush
    ErrorReset()
  end
  def LOGF
    ftr = ServiceGetFilter()
    if ftr.LinkGetTargetIdent() == "transFilter" then
      ReadForward(ftr)
    else
      @fh = File.new(ReadC(), "a")
    end
    SendRETURN()
  end
  def EXIT
    Process.exit!(1)
  end
  def Event
    if @itms.length == 0 then
      ErrorSetCONTINUE()
    else
      begin
	ftr = ServiceGetFilter()
        ftr.LinkConnect
	ReadForward(ftr,@itms[0])
      rescue Exception => ex
        ErrorSet(ex)
        if ErrorIsEXIT() then
          ErrorReset()
          return
        else
          ErrorWrite()
	end
      end
      @itms.shift
    end
  end
  def FilterIn
    @itms.push(ReadDUMP())
    SendRETURN()
  end
end
FactoryDefault("transFilter", Filter4)
srv = Filter4.new
begin
  srv.LinkCreate($0,ARGV)
  srv.ProcessEvent(MqS::WAIT_FOREVER)
rescue SignalException => ex
  # ignore
rescue Exception => ex
  srv.ErrorSet(ex)
ensure
  srv.Exit()
end

